### ISSUE TITLE
Is dynamic a member of dart:core?

### ISSUE THREAD

**Issue by peter-ahe-google** [dart-lang/sdk#29125]
According to the test `tests/language/import_core_prefix_test.dart`, `dynamic` is a member of `dart:core`.

According to the test `tests/language/dynamic_prefix_core_test.dart`, `dynamic` isn't a member of `dart:core`.

Which is it?

Please, this is a question about Dart 1.0, not about what changes are planned or might happen.

----

TODO (added Dec'17 by @anders-sandholm):
- [x] Update spec to say that `dynamic` is exported
- [x] Fix the test tests/language/dynamic_prefix_core_test.dart where dynamic isn't a member of dart:core.

Edit by @eernstg Feb 2018: The correction to the test file is now a separate issue (#32060), and consequently the second check box above was checked.

**Comment by peter-ahe-google** [comment]
@floitschG @eernstg @lrhn

**Comment by peter-ahe-google** [comment]
The Dart VM thinks that `dynamic` is a member of `dart:core`:

```
$ ./xcodebuild/ReleaseX64/dart xcodebuild/ReleaseX64/generated_tests/language/dynamic_prefix_core_test_none.dart
Unhandled exception:
NoSuchMethodError: No top-level getter 'dynamic' declared.
Receiver: top-level
Tried calling: dynamic
#0      NoSuchMethodError._throwNew (dart:core-patch/errors_patch.dart:194)
#1      main (file:///Users/ahe/w/dart-sdk/sdk/xcodebuild/ReleaseX64/generated_tests/language/dynamic_prefix_core_test_none.dart:12:17)
#2      _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:261)
#3      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:148)
$ ./xcodebuild/ReleaseX64/dart tests/language/import_core_prefix_test.dart
Instance of 'Object'
Instance of 'Object'
$ ./xcodebuild/ReleaseX64/dart -c tests/language/import_core_prefix_test.dart
Instance of 'Object'
Instance of 'Object'
$ ./xcodebuild/ReleaseX64/dart --version
Dart VM version: 1.23.0-edge.1c47632b5667027cb285a2a44d703a61e1f80071 (Tue Mar 21 10:54:44 2017) on "macos_x64"
```

Unfortunately, dart2js disagrees:

```
$ ./sdk/bin/dart2js --analyze-only xcodebuild/ReleaseX64/generated_tests/language/dynamic_prefix_core_test_none.dart
$ ./sdk/bin/dart2js --analyze-only tests/language/import_core_prefix_test.dart
tests/language/import_core_prefix_test.dart:28:3:
Warning: Cannot resolve type 'mycore.dynamic'.
  mycore.dynamic d = null;
  ^^^^^^^^^^^^^^
```

**Comment by kmillikin** [comment]
It's simultaneously both a member and not a member.  Until you open the box.

**Comment by peter-ahe-google** [comment]
The analyzer agrees with dart2js:

```
$ ./xcodebuild/ReleaseX64/dart-sdk/bin/dartanalyzer tests/language/import_core_prefix_test.dart
Analyzing tests/language/import_core_prefix_test.dart...
[warning] Undefined class 'mycore.dynamic' at tests/language/import_core_prefix_test.dart:28:3 (undefined_class).
[hint] The value of the field '_lat' isn't used at tests/language/import_core_prefix_test.dart:18:7 (unused_field).
[hint] The value of the local variable 'test' isn't used at tests/language/import_core_prefix_test.dart:24:7 (unused_local_variable).
[hint] The value of the local variable 'boolval' isn't used at tests/language/import_core_prefix_test.dart:25:15 (unused_local_variable).
[hint] The value of the local variable 'variable' isn't used at tests/language/import_core_prefix_test.dart:26:14 (unused_local_variable).
[hint] The value of the local variable 'value' isn't used at tests/language/import_core_prefix_test.dart:27:14 (unused_local_variable).
[hint] The value of the local variable 'd' isn't used at tests/language/import_core_prefix_test.dart:28:18 (unused_local_variable).
1 warning and 6 hints found.
```

**Comment by peter-ahe-google** [comment]
See also issue #12478.

**Comment by eernstg** [comment]
As far as I can see, the spec does not specify where `dynamic` must be declared, or how (in some magic way) it would otherwise be possible to perform a lookup and find it. I'd assume that it will work to (1) declare that it is defined and exported by 'dart:core', and (2) relying on the fact that it is a built-in identifier, enforce that it cannot be the name of a user-defined type, but it can be the name of non-type entities such as local variables.

We will then have to handle twisted examples where `dynamic` as a type is shadowed by a parameter declaration etc., but this is the most consistent treatment as far as I can see.

**Comment by peter-ahe-google** [comment]
FWIW, I've implemented having `dynamic` as a member of `dart:core`, and I can probably also implement the opposite.

**Comment by rmacnak-google** [comment]
In the VM's implementation, dynamic is a member of dart:core, but I've always considered this to be a bug. dynamic should always be available like true, false and null.

**Comment by lrhn** [comment]
The *dynamic* type is not a class, it is not declared anywhere, not even in `dart:core`.
The built-in identifier `dynamic`, when used as a type, refers to the *dynamic* type. That's not due to identifier resolution, it's by spec mandate.

The specification isn't very clear on this, so we have wiggle-room, and the above is my interpretation.

The statement that the name `dynamic` denotes a *Type* object is suspicious. Built-in identifiers can be declared as, say, local variables, so:
```dart
foo() {
  var dynamic = 42;
  print(dynamic);  // 42 or Type object for dynamic? Should be 42.
  bar(dynamic x) => x;  // This refers to the dynamic type, not as an identifier, but as a built-in identifier.
}
```

**Comment by peter-ahe-google** [comment]
I tend to agree with @rmacnak-google: `dynamic` should behave in the same way `true` and `false` do.

Even if it's decided that `dynamic` should be a reserved word, Fasta should continue to implement a weird rules for it for optimal backwards compatibility. What I mean by that, is that Fasta should accept @lrhn's example (even if `dynamic` becomes a reserved word) and generate exactly the same program as today, but it will also say that this program contains an error.

**Comment by eernstg** [comment]
The fact that `mycore.true` won't even parse prevents `true` (and `false`, and other reserved words) from being in the namespace exported by `dart:core`. So we _have to_ invent a different mechanism than scoped lookups in order to give a semantics to these words. This works quite well, every tool just "knows `true` already" whenever it is encountered.

I also think this makes sense psychologically: We don't want to shake the developer's world, so `true`, `42`, and other literals have a completely fixed meaning that nobody can undermine by importing an over-smart library.

`dynamic` is not a reserved word but a built-in identifier, so we're allowed to declare non-types with that name. Note that @lrhn's example is an error according to the analyzer ("Local variable 'dynamic' can't be referenced before it is declared", at the parameter to `bar` ;-) so we're not talking about a heavily used feature here.

I think one rather clean approach would be to make `dynamic` a reserved word, i.e., it cannot occur in constructs like `mycore.dynamic`, and it cannot be the name of a local variable or function etc.  In return, we ensure that it gets the same level of protection as `true` and `false`. I know, the 1st amendment to the Dart constitution says that our reserved words shall come straight out of JavaScript, but I don't see the need to insist on that today.

If we keep `dynamic` as a built-in identifier then there is no way (that I can see) to avoid talking about lookup: In some cases `dynamic` is the name of a variable, and we can only discover this by looking it up according to the scope rules. I'm sure we could come up with some rules such that it would be subject to lookup in some positions (`var y = dynamic;`, `dynamic.foo()`, and `foo(mycore.dynamic x) => x;`) and not in others (`dynamic z = mycore.int`), but I'd prefer if we could avoid making twisted rules just because a couple of programs out there are using variables named `dynamic`.  I'm not worried about the effort needed to write a specification for those twisted rules, I'm worried about the wasted brain cycles for every developer who ever needs to know about them.

**Comment by eernstg** [comment]
In [this message on dart-discuss](https://groups.google.com/a/google.com/d/msg/dart-discuss/A-TvO9b90AQ/Z5k144rcBwAJ), @lrhn concludes that we already have a resolution:

> dart2js might be odd, but it's also correct [when it says that `dynamic` is not exported from core]

I agree that the spec does not explicitly specify that `dynamic` is exported from core. However, the spec actually has the following (which has been there, unchanged, since Oct 2013 when the spec was added to this repo):

> References to otherwise unspecified names of program entities (such as classes or functions)
> are interpreted as the names of members of the Dart core library.

I find it difficult to maintain that `dynamic` used as/in an expression or a type annotation isn't the 'name of a program entity', so you could actually claim that we have a resolution, except that it's the opposite one: `dynamic` _is_ exported from core. ;-)

Nevertheless, I'd recommend that we clarify the situation such that `dynamic` is not exported from core, and it is not subject to lexical lookups (so we make it a reserved word), and then it shouldn't be hard to make every tool follow those rules.

**Comment by eernstg** [comment]
(Yes, I know this breaks every program which declares a local variable, a parameter, etc. named `dynamic`, but unless that's a serious practical problem I think we'll be better off without it).

**Comment by lrhn** [comment]
True, the spec isn't particularly clear about what `dynamic` is. 
I'm saying that it is not exported by `dart:core` because it is not *declared* in dart:core.

We could change that, but I don't think we'll help anybody by doing so.
Before going overboard and declaring `dynamic` a reserved word, we might want to first consider if we 
 want to rename the type to something shorter for Dart 2, or if we *want* to make it an identifier exported by dart:core.

**Comment by peter-ahe-google** [comment]
Would it be possible to discuss this on a language team meeting and get back to this bug once the discussion has been resolved?

**Comment by leafpetersen** [comment]
Where are we on this?

**Comment by kevmoo** [comment]
Ping? @sigmundch ? @rakudrama ?

**Comment by sigmundch** [comment]
we suppressed the hint in dart2js back in 98547dc, we are waiting on language resolution at this moment.

**Comment by lrhn** [comment]
Language decision is that `dynamic` is a type exported by `dart:core`. (It's a non-class type, similar to `FutureOr` exported by `dart:async`, but the *name* is treated as an identifier exported by `dart:core`).

**Comment by kevmoo** [comment]
Yay! Glad this is decided!

**Comment by bwilkerson** [comment]
> Language decision is that `dynamic` is a type exported by `dart:core`.

Will it be declared in `dart:core`?

**Comment by rmacnak-google** [comment]
And what about `void`?

**Comment by lrhn** [comment]
The name `dynamic` is in the export scope of `dart:core`, and it denotes the dynamic type. That is basically all there is to say about it.

The name `dynamic` is a built-in identifier, so when it's used, the specification just treats it as an identifier. That means that it should exist in some scope, so we pick the scope that is always imported unless you actively try to avoid that. Since it's a built-in identifier, you cannot declare another type with that name (but you can shadow it by declaring a non-type named `dynamic` ... but please don't!).

There isn't currently any declaration of `dynamic` in the library source code. We can't write one since there is no syntax for non-class/non-function type declarations. We could fake it, like we currently do for `FutureOr`, by having a class declaration and then documenting that it's not actually a class type anyway, but I don't see a good reason for doing that for `dynamic`. The implementations need to do magic in either case, by adding the "magical" type declaration, and having to remove an existing class declaration too is just more work. Whether we retain the declaration for `FutureOr` or not isn't something I have thought deeply about, since it won't have any practical effect on users. It's just a placeholder for documentation.
In this case, we'll just do whatever is easier for the tools, it has no end-user impact whether there is a line in the source code or not.

The name `void` is a reserved word, not an identifier. As such, it's never looked up in any scope (there is no part of the spec saying to do so, that's something you do to identifiers), so putting it in a scope is unnecessary. The syntactic construct `void` directly denotes the void type (which will be a real type in Dart 2, not just something you write instead of a return type on a function).
We can change `void` to be a built-in identifier, and did plan to at some point, but the return on that effort is probably zero at best. At least for now, `void` is still a reserved word.

(Maybe, eventually, we can add some syntax like `external type dynamic;` and `external type FutureOr<T>;` to the grammar, just so we have some source ocde, but it won't actually *change* anything, so it seems like wasted effort).

**Comment by eernstg** [comment]
No milestone now: This is independent of Dart 2.

Current status: The spec says that

> References to otherwise unspecified names of program entities (such as
> classes or functions) are interpreted as the names of members of the Dart
> core library.

All such names would have to be exported in order to be relevant in the spec, hence the given decision (`dynamic` is exported by 'dart:core') is already what the spec says. In case there is a need to say the same thing in a different way, please create a new issue making that request.

The test file correction has been transferred to a new issue (#32060), because it does not fit the area labels on this issue. That issue refers to this one as background material.

With that, we have resolved the area-language and the area-specification part which are the labels of this issue, and the test file part has been separated out into a new issue with an area-test label.

I believe this issue is then resolved; closing.
